### ISSUE TITLE
Question: Why stateless functions aren't referentially transparent, if they use defaultProps

### ISSUE THREAD

**Issue by tricoder42** [acdlite/recompose#215]
There're actually two ways how to set defaultProps for stateless components:

```js
const A = (props = defaultValues) => {};

const B = (props) => {};
B.defaultProps = defaultValues;
```

Looking at the code of [isReferentiallyTransparentFunctionComponent](https://github.com/acdlite/recompose/blob/master/src/packages/recompose/isReferentiallyTransparentFunctionComponent.js) function, the component `A` would be incorrectly considered as referentially transparent.

I probably don't understand the referential transparency correctly, but why stateless functions aren't referentially transparent, if they use defaultProps? I've also managed to find this commit https://github.com/acdlite/recompose/commit/122477e00d8aa2aa9ddb9667100fb41b74f53288, but there's no explanation...

**Comment by istarkov** [comment]
The problem with B that you can't just replace it with function call in HOC
this  `const C = (props) => B(props)` will give incorrect results as B default values will be never used.
but this  `const C = (props) => A(props)` will work as expected.

**Comment by wuct** [comment]
@tricoder42 I have created [a demo](http://www.webpackbin.com/VkxQpZO8Z) to show this behavior. Hope this could help.

**Comment by tricoder42** [comment]
Thank you both for explanation. Makes perfectly sense!

TL;DR: Using default parameters instead of defaultProps attribute is in general better, because it doesn't rely on React `createElement`, right?

**Comment by tricoder42** [comment]
I've just realized that using default value for `props` parameter doesn't work at all. React always provide `props` parameter, even if it's just empty object, so `defaultValues` constant in my example is never used.

As @wuct used in his example, default values must be used in object destructuring to make it work :)

```js
const Bad = (props = {a: 1}) => {};
const Good = ({a = 1}) => {};
```

**Comment by wuct** [comment]
I am going to close this issue. Please feel free to reopen it if you have further questions 😄
